### PR TITLE
matching logic prover support for list segments and overall improvements; more examples

### PR DESCRIPTION
**Pull request by andreistefanescu**

### CODE REVIEW

**Comment by andreistefanescu** [comment]
Jenkins: test this please

**Comment by andreistefanescu** [comment]
@yilongli @grosu @dwightguth please review.

File: src/main/java/org/kframework/backend/java/symbolic/SymbolicUnifier.java [hunk 1]
```diff
@@ -302,109 +329,171 @@ public boolean unifyList(BuiltinList list, BuiltinList otherList, boolean addUnc
...
         }
-        for (Term key : entries.keySet()) {
-            if (!commonKeys.contains(key)) {
-                remainingEntries.put(key, entries.get(key));
+
+        for (BuiltinMap foldedMap : foldedMaps) {
+            SymbolicConstraint constraint = new SymbolicConstraint(fConstraint.termContext());
+            constraint.add(foldedMap, otherMap);
+            constraint.simplify();
+            constraint.simplify();
```
Comment (yilongli) [inline src/main/java/org/kframework/backend/java/symbolic/SymbolicUnifier.java:362]: simplify twice?

**Comment by yilongli** [comment]
Overall, lgtm. My only concern is that `SymbolicUnifier`, `SymbolicConstraint` and `SymbolicRewriter` are getting more and more complex.

File: src/main/java/org/kframework/backend/java/symbolic/SymbolicUnifier.java (hunk 1 above)
Comment (andreistefanescu) [inline src/main/java/org/kframework/backend/java/symbolic/SymbolicUnifier.java:362]: yeah...I am having problems propagating the changes from map simplifications.

**Comment by andreistefanescu** [comment]
Jenkins: test this please

**Comment by grosu** [comment]
lgtm.  I think things will get a lot simpler with the new KORE/KAST.  There should be fewer cases, at least.

Grigore

File: src/main/java/org/kframework/backend/java/util/Z3Wrapper.java [hunk 2]
```diff
@@ -16,36 +17,42 @@
...
         return Z3_WRAPPER;
     }
 
     public final String SMT_PRELUDE;
 
-    public Z3Wrapper() {
+    public Z3Wrapper(Context context) {
         String s = "";
         try {
-            s = new String(Files.toByteArray(new File(JarInfo.getZ3PreludePath())));
```
Comment (yilongli) [inline src/main/java/org/kframework/backend/java/util/Z3Wrapper.java:31]: remove `JarInfo.getZ3PreludePath()` if it is not used anymore?

**Comment by yilongli** [comment]
Jenkins: retest this please

**Comment by andreistefanescu** [comment]
Jenkins: test this please

File: pom.xml [hunk 3]
```diff
@@ -173,6 +173,14 @@
       <classifier>${native.classifier}</classifier>
     </dependency>
     <dependency>
+      <groupId>com.microsoft.z3</groupId>
```
Comment (dwightguth) [inline pom.xml:176]: do we really need both the binary and the jar?

File: src/main/java/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java [hunk 4]
```diff
@@ -658,13 +658,15 @@ private static Definition evaluateDefinition(GlobalContext globalContext) {
         Iterable<Rule> rules = Iterables.concat(
                 definition.rules(),
                 definition.macros(),
-                definition.patternRules().values());
+                definition.patternRules().values(),
+                definition.patternFoldingRules());
```
Comment (dwightguth) [inline src/main/java/org/kframework/backend/java/symbolic/KILtoBackendJavaKILTransformer.java:662]: do we need to have some kind of group collection of rules the same way we do with baseTerms? or do you think that the issues that led to that generalization being desirable are not in play here?

**Pull request merged**